src/app/filelogger.cpp

```diff
@@ -31,32 +31,100 @@
 #include <chrono>
 
 #include <QDateTime>
+#include <QDebug>
 #include <QDir>
+#include <QHash>
 #include <QTextStream>
-#include <QVector>
 
 #include "base/global.h"
 #include "base/logger.h"
 #include "base/utils/fs.h"
+#include "base/utils/gzip.h"
 
 namespace
 {
     const std::chrono::seconds FLUSH_INTERVAL {2};
+
+    bool isObsolete(const QFileInfo &info, FileLogger::FileLogAgeType ageType, int age)
+    {
+        QDateTime modificationDate = info.lastModified();
+        switch (ageType)
+        {
+        case FileLogger::DAYS:
+            modificationDate = modificationDate.addDays(age);
+            break;
+        case FileLogger::MONTHS:
+            modificationDate = modificationDate.addMonths(age);
+            break;
+        default:
+            modificationDate = modificationDate.addYears(age);
+        }
+        return modificationDate <= QDateTime::currentDateTime();
+    }
+
+    bool compressBackupFile(const Path &sourcePath, const Path &destPath, int level, QString &msg)
+    {
+        QFile source {sourcePath.data()};
+        QFileInfo info {source};
+        const QDateTime atime = info.lastRead();
+        const QDateTime mtime = info.lastModified();
+        // It seems the created time can't be modified on UNIX.
+        const QDateTime ctime = info.birthTime();
+        const QDateTime mctime = info.metadataChangeTime();
+
+        if (!source.open(QIODevice::ReadOnly))
+        {
+            msg = FileLogger::tr("Can't open %1!").arg(sourcePath.data());
+            return false;
+        }
+
+        bool ok = false;
+        const QByteArray data = Utils::Gzip::compress(source.readAll(), level, &ok);
```

Comment:
Just pointing out. Not sure what is the possible upper size limit of the log file, if the size is big (> 100 MB I would guess) you are definitively going to see bad allocation bug reports coming in.

Reply:
Streaming should be better. Although I'm not sure `QDataStream` is suitable here, AFAIK it is not a pure streaming interface and it will inject its own data/serialization to the stream. Also don't have time to comment on the code details.

Reply:
> AFAIK it is not a pure streaming interface and it will inject its own data/serialization to the stream.

Used `readRawData()/writeRawData()` methods are intended for [Reading and Writing Raw Binary Data](https://doc.qt.io/qt-6/qdatastream.html#raw).

Reply:
> I added a function to compress files by QDataStream. What do you think about this? https://github.com/brvphoenix/qBittorrent/blob/529ce9178db16f3bf0af3ca8cb8e3180521998dc/src/base/utils/gzip.cpp#L44-L97

Could you provide `decompress()` counterpart too? I would like to test them.

Reply:
I also perceive `QDataStream` as some kind of auxiliary tool, and I would prefer to see `QIODevice` in the interface of `compress()`/`decompress()` methods.

Reply:
There is `QBuffer` adaptor class to use `QByteArray` as `QIODevice`.